packages/arb-ts/src/lib/message/L1Transaction.ts

```diff
@@ -0,0 +1,213 @@
+    this.to = tx.to
+    this.from = tx.from
+    this.contractAddress = tx.contractAddress
+    this.transactionIndex = tx.transactionIndex
+    this.root = tx.root
+    this.gasUsed = tx.gasUsed
+    this.logsBloom = tx.logsBloom
+    this.blockHash = tx.blockHash
+    this.transactionHash = tx.transactionHash
+    this.logs = tx.logs
+    this.blockNumber = tx.blockNumber
+    this.confirmations = tx.confirmations
+    this.cumulativeGasUsed = tx.cumulativeGasUsed
+    this.effectiveGasPrice = tx.effectiveGasPrice
+    this.byzantium = tx.byzantium
+    this.type = tx.type
+    this.status = tx.status
+  }
+
+  /**
+   * Get the numbers of any messages created by this transaction
+   * @returns
+   */
+  public getMessageNumbers(): BigNumber[] {
+    const iface = Inbox__factory.createInterface()
+    const messageDelivered = iface.getEvent('InboxMessageDelivered')
+    const messageDeliveredFromOrigin = iface.getEvent(
+      'InboxMessageDeliveredFromOrigin'
+    )
+    const eventTopics = {
+      InboxMessageDelivered: iface.getEventTopic(messageDelivered),
+      InboxMessageDeliveredFromOrigin: iface.getEventTopic(
+        messageDeliveredFromOrigin
+      ),
+    }
+    const logs = this.logs.filter(
+      log =>
+        log.topics[0] === eventTopics.InboxMessageDelivered ||
+        log.topics[0] === eventTopics.InboxMessageDeliveredFromOrigin
+    )
+    return logs.map(log => BigNumber.from(log.topics[1]))
+  }
+
+  /**
+   * Get any l1tol2 messages created by this transaction
+   * @param l2SignerOrProvider
+   */
+  public async getL1ToL2Messages<T extends SignerOrProvider>(
+    l2SignerOrProvider: T
+  ): Promise<L1ToL2MessageReaderOrWriter<T>[]>
+  public async getL1ToL2Messages<T extends SignerOrProvider>(
+    l2SignerOrProvider: T
+  ): Promise<L1ToL2MessageReader[] | L1ToL2MessageWriter[]> {
+    const provider = SignerProviderUtils.getProviderOrThrow(l2SignerOrProvider)
+
+    const chainID = (await provider.getNetwork()).chainId.toString()
+
+    const messageNumbers = this.getMessageNumbers()
+    if (!messageNumbers || messageNumbers.length === 0) return []
+
+    return messageNumbers.map((mn: BigNumber) => {
+      const ticketCreationHash = L1ToL2Message.calculateRetryableCreationId(
+        BigNumber.from(chainID),
+        mn
+      )
+      return L1ToL2Message.fromRetryableCreationId(
+        l2SignerOrProvider,
+        ticketCreationHash,
+        mn
+      )
+    })
+  }
+
+  /**
+   * Gets a single l1ToL2Message
+   * If the messageIndex is supplied the message at that index will be returned.
+   * If no messageIndex is supplied a message will be returned if this transaction only created one message
+   * All other cases throw an error
+   * @param l2SignerOrProvider
+   */
+  public async getL1ToL2Message<T extends SignerOrProvider>(
+    l2SignerOrProvider: T,
+    messageNumberIndex?: number
+  ): Promise<L1ToL2MessageReaderOrWriter<T>>
+  public async getL1ToL2Message<T extends SignerOrProvider>(
```

Comment:
Is this needed?
It doesn't actually have a performance advantage compared to `getL1ToL2Messages` 
Just adds some message count validation then returns the desired index

Reply:
Yeah, I kinda agree. @DZGoldman mentioned that this is the most normal way of consuming the messages, and so was useful to have.

Reply:
We never use this internally, and externally its an extra couple of lines of validation which we let the consumer decide on